### ISSUE TITLE
Request: PID Middleware

### ISSUE THREAD

**Issue by avluis** [CANBus-Triple/CANBus-Triple#17]
Hello there~
I like where the project is headed and the freedom provided by the CANBus Triple when it comes to how it lets you gather and process your data.
Unfortunately, I am also in a bind due to that - I'm a simple Arduino hobbyist with very low programming experience but with plenty of ideas when it comes to making my ride more convenient.
Just recently I finished major work with a custom made, [Arduino powered, heated seats](https://github.com/avluis/ArduinoHeatedSeatController) interface controller.
If you take the time to look over that, you should be able to tell that due to my lack of experience (self taught), I have avoided using classes, pointers and anything deemed advanced when it comes to C programming. Well, I could say it was not really necessary, but any program benefits from proper use of classes (I also believe in keeping it simple).

Anyways, I rambled on, sorry.
Back to the point:
I believe the CANBus Triple supports reading all sorts of data provided by the CAN network so it is only limited by the programmer - and so my request:
In the same way that you can request data from the CANBus Triple about your vehicle, is it possible to implement a PID Middleware that resolves common PID values directly on the CANBus Triple, which when queried will then enter a PID relay mode (polls and resolves PID data from the vehicle, ready to be sent to the requesting system) until requested or powered off.
The goal is to be able to recreate a common OBD-II PID Scanner (in this case with Bluetooth).
The app that is currently in the works could benefit directly from this as well as a way to tell the system is working with no additional programming from the moment that it is plugged into a vehicle by a new owner.

So, TL;DR
Implement a PID Middleware relay module for use with the desktop/mobile applications and/or whichever other way you can communicate with the CANBus Triple.

**Comment by tz1** [comment]
I actually have that on my "todo" list, elm327/329 emulation, but that doesn't do PIDs just would act like an adapter.

(Right now I'm waiting for some more direction and examples on using Bluetooth, and I'm in the middle of a relocation).

The actual OBD-II PID logic is simple, then only initial probes would need to determine 11 or 29 bit CAN and which modules respond (e.g. RPM can be returned from both the engine and transmission).
It would be fairly easy to make a Torque-o-matic, at least for the raw OBD-II data.

I already have a simple, raw diagnostic replay sketch (at my tz1 fork) that just sends what it sees on the bus via USB.  I need to add autobaud.  But the sketch is much simpler.

**Comment by etx** [comment]
This _kind of_ exists right now as the ServiceCall middleware. 

The CBT can hold 8 PIDs in the EEPROM and the ServiceCall middleware will listen for passive packets or dispatch packets requesting the data it wants. For every CAN message that comes in ServiceCall will look see if it matches one of the 8 PIDs in and if so will run the math on them and save the result to memory, and also send it to the bluetooth le module so it is retained in the GATT. 

I'm working on the app to make it easy to modify these. They use Xgauge format as described here:
http://www.scan-gauge.com/wp-content/uploads/XGaugeCoding.pdf

On first boot of the device defaults are added to the EEPROM memory. If you wanted to mess with this functionality right now easily you could edit the default settings in the firstBoot method here:
https://github.com/CANBus-Triple/CANBus-Triple/blob/master/avr/examples/CANBusTriple/Settings.h

Then flash the code to the CBT and send a 0x0104 command over serial to tell the CBT to write the eeprom with default values.

**Comment by avluis** [comment]
I will definitely check that functionality out right now, so thank you for the heads up.
